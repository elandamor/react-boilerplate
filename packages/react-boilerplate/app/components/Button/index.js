/**
 * Button
 * @type {Component} Presentational
 */

import React from 'react';
import PropTypes from 'prop-types';
// Styled-Components
import Wrapper from './styles';

const Button = ({ children }) => (
  <Wrapper>{children}</Wrapper>
);

Button.propTypes = {
  children: PropTypes.any.isRequired,
};

export default Button;
