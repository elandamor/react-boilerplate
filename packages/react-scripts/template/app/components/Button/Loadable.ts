/**
 *
 * Asynchronously loads the component for Button
 *
 */

import React from 'react';

export default React.lazy(() => import('./index'));
