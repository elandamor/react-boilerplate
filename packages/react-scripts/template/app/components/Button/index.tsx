import React, { MouseEvent, FC } from 'react';
import classNames from 'classnames';
// Styles
import Wrapper from './styles';

export interface IButtonProps {
  backgroundColor?: string;
  className?: string;
  disabled?: boolean;
  icon?: JSX.Element;
  iconOnly?: boolean;
  iconPosition?: string;
  iconSize?: number;
  ml?: number | string;
  mr?: number | string;
  mx?: number | string;
  onClick?: (event: MouseEvent<HTMLElement>) => void;
  outlined?: boolean;
  raised?: boolean;
  text?: string;
  textColor?: string;
  type?: 'button' | 'submit';
}

/**
 * @render react
 * @name Button component
 * @description Defines a clickable button.
 * @example
 * <Button text="Test" />
 */

const Button: FC<IButtonProps> = ({
  className,
  onClick: handleClick,
  text,
  ...rest
}) => (
  <Wrapper
    className={classNames('c-btn', className)}
    onClick={handleClick}
    {...rest}
  >
    {rest.icon && rest.iconPosition === 'left' && <i>{rest.icon}</i>}
    {!rest.iconOnly && <label>{text}</label>}
    {rest.icon && rest.iconPosition === 'right' && <i>{rest.icon}</i>}
  </Wrapper>
);

Button.defaultProps = {
  iconOnly: false,
  iconPosition: 'left',
  onClick: () => null,
  text: 'Button',
  type: 'button',
};

export default Button;
