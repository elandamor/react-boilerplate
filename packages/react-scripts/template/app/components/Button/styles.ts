import styled, { css } from 'styled-components';
import { space } from 'styled-system';
import { IButtonProps } from './index';
import { THEME } from '../../global-styles';

const Wrapper = styled.button`
  ${space};
  align-items: center;
  background: transparent;
  border: none;
  border-radius: ${THEME.space[1] / 2}px;
  cursor: ${({ disabled }: IButtonProps) => disabled ? 'not-allowed' : 'pointer'};
  display: flex;
  font-size: ${THEME.fontSizes[1]}px;
  justify-content: center;
  letter-spacing: .0892857143em;
  min-height: ${THEME.space[4] + (THEME.space[1] / 2)}px;
  min-width: ${THEME.space[8]}px;
  opacity: ${({ disabled, raised }: IButtonProps) => raised && disabled ? '0.38' : '1'};
  outline: none;
  padding: ${THEME.space[1]}px;
  position: relative;
  text-transform: uppercase;
  z-index: 1;

  i {
    pointer-events: none;
    z-index: 1;
  }

  i, i > svg {
    height: ${({ iconSize }: IButtonProps) => iconSize ? iconSize : '18'}px;
    width: ${({ iconSize }: IButtonProps) => iconSize ? iconSize : '18'}px;
  }

  label {
    pointer-events: none;
    z-index: 1;

    ${({ icon, iconPosition }: IButtonProps) =>
      icon && iconPosition !== 'right' && css`
        margin-left: ${THEME.space[1]}px;
      `
    };

    ${({ icon, iconPosition }: IButtonProps) =>
      icon && iconPosition === 'right' && css`
        margin-right: ${THEME.space[1]}px;
      `
    };
  }

  &:after, &:before {
    content: '';
    height: 100%;
    position: absolute;
    left: 0;
    top: 0;
    width: 100%;
    z-index: 0;
  }

  &:not([disabled]) {
    &:focus {
      &:before {
        ${({ raised }: IButtonProps) => !raised && css`
          background-color: ${THEME.colors.blacks[3]};
        `};
      }

      &:after {
        ${({ raised }: IButtonProps) => raised && css`
          background-color: ${THEME.colors.whites[3]};
        `};
      }
    }

    &:hover {
      &:before {
        ${({ raised }: IButtonProps) => !raised && css`
          background-color: ${THEME.colors.blacks[2]};
        `};
      }

      &:after {
        ${({ raised }: IButtonProps) => raised && css`
          background-color: ${THEME.colors.blacks[2]};
        `};
      }
    }
  }

  ${({ outlined }: IButtonProps) => outlined && css`
    border: ${THEME.borders[1]} ${THEME.colors.blacks[3]};
    padding: ${THEME.space[1]}px ${THEME.space[2]}px;
  `};

  ${({ backgroundColor, raised, textColor }: IButtonProps) => raised && css`
    background-color: ${backgroundColor || THEME.colors.primary};
    box-shadow:
      0 3px 1px -2px ${THEME.colors.blacks[4]},
      0 2px 2px 0 ${THEME.colors.blacks[3]},
      0 1px 5px 0 ${THEME.colors.blacks[3]};
    color: ${textColor || THEME.colors.white};
    padding: ${THEME.space[1]}px ${THEME.space[2]}px;
  `};

  ${({ icon, iconPosition }: IButtonProps) =>
    icon && iconPosition !== 'right'  && css`
      padding-left: ${THEME.space[1] / 2}px;
    `
  };

  ${({ icon, iconPosition }: IButtonProps) =>
    icon && iconPosition === 'right'  && css`
      padding-right: ${THEME.space[1] / 2}px;
    `
  };

  ${({ icon, iconPosition, outlined, raised }: IButtonProps) =>
    icon && iconPosition !== 'right' && (outlined || raised) && css`
      padding-left: ${THEME.space[1] + (THEME.space[1] / 2)}px;
    `
  };

  ${({ icon, iconPosition, outlined, raised }: IButtonProps) =>
    icon && iconPosition === 'right' && (outlined || raised) && css`
      padding-right: ${THEME.space[1] + (THEME.space[1] / 2)}px;
    `
  };

  ${({ iconOnly }: IButtonProps) => iconOnly && css`
      min-width: 40px;
      padding: 0;
    `
  };
`;

export default Wrapper;
