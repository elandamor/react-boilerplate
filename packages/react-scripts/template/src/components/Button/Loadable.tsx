/**
 *
 * Asynchronously loads the component for Button
 *
 */

import React from 'react';
import { Loadable } from '../../utils';
import LoadingBar from '../LoadingBar';

export default Loadable(() => import('./index'), {
  fallback: <LoadingBar />,
});
