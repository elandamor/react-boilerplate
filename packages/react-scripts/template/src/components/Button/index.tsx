import React, { FC } from 'react';
import { StyledSystemProps } from 'styled-system';
// Styles
import Wrapper from './styles';

export interface IButtonProps extends StyledSystemProps {
  ariaLabel?: string;
  disabled?: boolean;
  icon?: React.ReactNode;
  iconOnly?: boolean;
  iconSize?: number;
  onClick?: (event: React.MouseEvent<HTMLButtonElement>) => void;
  text?: string;
  type?: 'button' | 'submit';
  // Variants
  flat?: boolean;
  outlined?: boolean;
  primary?: boolean;
  raised?: boolean;
  secondary?: boolean;
  variant?: 'flat' | 'outlined' | 'primary' | 'raised' | 'secondary' | 'text';
}

/**
 * @render react
 * @name Button component
 * @description Defines a clickable button.
 * @example
 * <Button text="Test" />
 */

const Button: FC<IButtonProps> = ({ text, ...rest }) => (
  // @ts-ignore
  <Wrapper {...rest}>
    {rest.icon && <i>{rest.icon}</i>}
    {!rest.iconOnly && <label>{text}</label>}
  </Wrapper>
);

Button.defaultProps = {
  ariaLabel: 'Button',
  bg: 'transparent',
  color: 'text',
  disabled: false,
  iconOnly: false,
  iconSize: 18,
  minWidth: 64,
  onClick: () => null,
  text: 'Button',
  type: 'button',
  variant: 'text',
};

export default Button;
