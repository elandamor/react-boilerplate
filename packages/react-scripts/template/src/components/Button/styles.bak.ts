import styled, { css } from 'styled-components';
import { color, space } from 'styled-system';
import { IButtonProps } from './index';
import theme from '../../theme';

const Wrapper = styled.button<IButtonProps>`
  ${color};
  ${space};
  align-items: center;
  background: transparent;
  border: none;
  border-radius: ${theme.space[1] / 2}px;
  cursor: ${({ disabled }) => disabled ? 'not-allowed' : 'pointer'};
  display: flex;
  font-size: ${theme.fontSizes[2]}px;
  font-weight: ${theme.fontWeights[4]};
  justify-content: center;
  letter-spacing: .0892857143em;
  min-height: ${theme.space[4] + (theme.space[1] / 2)}px;
  min-width: ${theme.space[8]}px;
  opacity: ${({ disabled, raised }) => raised && disabled ? '0.38' : '1'};
  outline: none;
  padding: ${theme.space[1]}px;
  position: relative;
  text-transform: uppercase;
  z-index: 1;

  i {
    pointer-events: none;
    z-index: 1;
  }

  i, i > svg {
    height: ${({ iconSize }) => iconSize ? iconSize : '18'}px;
    width: ${({ iconSize }) => iconSize ? iconSize : '18'}px;
  }

  label {
    pointer-events: none;
    z-index: 1;

    ${({ icon, iconPosition }) =>
      icon && iconPosition !== 'right' && css`
        margin-left: ${theme.space[1]}px;
      `
    };

    ${({ icon, iconPosition }) =>
      icon && iconPosition === 'right' && css`
        margin-right: ${theme.space[1]}px;
      `
    };
  }

  &:after, &:before {
    content: '';
    height: 100%;
    position: absolute;
    left: 0;
    top: 0;
    width: 100%;
    z-index: 0;
  }

  &:not([disabled]) {
    &:focus {
      &:before {
        ${({ raised }) => !raised && css`
          background-color: ${({ theme }) => theme.isDark
            ? theme.colors.whites[3] : theme.colors.blacks[3]};
        `};
      }

      &:after {
        ${({ raised }) => raised && css`
          background-color: ${({ theme }) => theme.isDark
            ? theme.colors.blacks[3] : theme.colors.whites[3]};
        `};
      }
    }

    &:hover {
      &:before {
        ${({ raised }) => !raised && css`
          background-color: ${({ theme }) => theme.isDark
            ? theme.colors.whites[2] : theme.colors.blacks[2]};
        `};
      }

      &:after {
        ${({ raised }) => raised && css`
          background-color: ${({ theme }) => theme.isDark
            ? theme.colors.whites[2] : theme.colors.blacks[2]};
        `};
      }
    }
  }

  ${({ outlined }) => outlined && css`
    border: ${theme.borders[1]} ${({ theme }) => theme.isDark
      ? theme.colors.whites[1] : theme.colors.blacks[1]};
    padding: ${theme.space[1]}px ${theme.space[2]}px;
  `};

  ${({ backgroundColor, raised, textColor }) => raised && css`
    background-color: ${backgroundColor || theme.colors.primary};
    box-shadow:
      0 3px 1px -2px ${theme.colors.blacks[4]},
      0 2px 2px 0 ${theme.colors.blacks[3]},
      0 1px 5px 0 ${theme.colors.blacks[3]};
    color: ${textColor || theme.colors.white};
    padding: ${theme.space[1]}px ${theme.space[2]}px;
  `};

  ${({ icon, iconPosition }) =>
    icon && iconPosition !== 'right'  && css`
      padding-left: ${theme.space[1] / 2}px;
    `
  };

  ${({ icon, iconPosition }) =>
    icon && iconPosition === 'right'  && css`
      padding-right: ${theme.space[1] / 2}px;
    `
  };

  ${({ icon, iconPosition, outlined, raised }) =>
    icon && iconPosition !== 'right' && (outlined || raised) && css`
      padding-left: ${theme.space[1] + (theme.space[1] / 2)}px;
    `
  };

  ${({ icon, iconPosition, outlined, raised }) =>
    icon && iconPosition === 'right' && (outlined || raised) && css`
      padding-right: ${theme.space[1] + (theme.space[1] / 2)}px;
    `
  };

  ${({ iconOnly }) => iconOnly && css`
      min-width: 40px;
      padding: 0;
    `
  };
`;

export default Wrapper;
