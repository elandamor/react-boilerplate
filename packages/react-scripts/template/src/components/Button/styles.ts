import styled, { css } from 'styled-components';
import { borderRadius, color, minWidth, size, space } from 'styled-system';
import { IButtonProps } from './index';
import theme from '../../theme';

const Wrapper = styled.button<IButtonProps>`
  align-items: center;
  background: transparent;
  border: none;
  cursor: ${({ disabled }) => disabled ? 'not-allowed' : 'pointer'};
  display: flex;
  font-size: ${theme.fontSizes[2]}px;
  font-weight: ${theme.fontWeights[4]};
  justify-content: center;
  letter-spacing: .0892857143em;
  min-height: ${theme.space[4] + (theme.space[1] / 2)}px;
  opacity: ${({ disabled, raised }) => raised && disabled ? '0.38' : '1'};
  outline: none;
  overflow: hidden;
  padding: ${theme.space[1]}px;
  position: relative;
  text-transform: uppercase;
  z-index: 1;
  ${borderRadius};
  ${color};
  ${minWidth};
  ${size};
  ${space};

  i {
    pointer-events: none;
    z-index: 1;
  }

  i, i > svg {
    height: ${({ iconSize }) => iconSize ? iconSize : '18'}px;
    width: ${({ iconSize }) => iconSize ? iconSize : '18'}px;
  }

  &:after, &:before {
    content: '';
    height: 100%;
    position: absolute;
    left: 0;
    top: 0;
    width: 100%;
    z-index: 0;
  }

  &:active {
    transform: scale(0.95);
  }

  &:not([disabled]) {
    &:focus {
      &:before {
        ${({ raised }) => !raised && css`
          background-color: ${({ theme }) => theme.isDark
            ? theme.colors.whites[3] : theme.colors.blacks[3]};
        `};
      }

      &:after {
        ${({ raised }) => raised && css`
          background-color: ${({ theme }) => theme.isDark
            ? theme.colors.blacks[3] : theme.colors.whites[3]};
        `};
      }
    }

    &:hover {
      &:before {
        ${({ raised }) => !raised && css`
          background-color: ${({ theme }) => theme.isDark
            ? theme.colors.whites[2] : theme.colors.blacks[2]};
        `};
      }

      &:after {
        ${({ raised }) => raised && css`
          background-color: ${({ theme }) => theme.isDark
            ? theme.colors.whites[2] : theme.colors.blacks[2]};
        `};
      }
    }
  }
`;

export default Wrapper;
