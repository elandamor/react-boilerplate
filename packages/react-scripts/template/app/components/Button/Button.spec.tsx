// Button.spec.tsx
import * as React from 'react';
import { cleanup, fireEvent, render } from 'react-testing-library';
import 'jest-dom/extend-expect';

import Button from './index';

afterEach(cleanup);

describe('Button', () => {
  it('should render with text', () => {
    const props = {
      text: 'Test',
    };

    const { getByText } = render(<Button {...props} />);

    const testButton = getByText(props.text);

    expect(testButton).toBeDefined();
  });

  it('should handle onClick event', () => {
    function Counter() {
      const [count, setCount] = React.useState(0);

      const props = {
        onClick: () => setCount(count + 1),
        text: count.toString(),
      };

      return <Button {...props} />;
    }

    const {
      container: { firstChild: buttonNode },
    } = render(<Counter />);

    // @ts-ignore
    fireEvent.click(buttonNode);
    expect(buttonNode).toHaveTextContent('1');
  });
});
