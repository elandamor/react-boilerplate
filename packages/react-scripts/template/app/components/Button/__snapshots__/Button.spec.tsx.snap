// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> should display a button 1`] = `
<button
  className="c-btn "
  onClick={[Function]}
/>
`;
